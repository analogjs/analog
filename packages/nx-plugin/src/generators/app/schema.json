{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "app",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed"
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    }
  },
  "required": ["name"]
}
