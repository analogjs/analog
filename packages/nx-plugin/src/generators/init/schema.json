{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "page",
  "title": "Adds an Analog configuration",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The project that will be updated.",
      "x-prompt": "Which project would you like to update?",
      "$default": {
        "$source": "projectName"
      },
      "x-dropdown": "projects"
    },
    "vitest": {
      "type": "boolean",
      "description": "Add testing with Vitest",
      "x-prompt": "Would you like to configure unit testing with Vitest?",
      "default": false
    }
  },
  "required": ["project"]
}
