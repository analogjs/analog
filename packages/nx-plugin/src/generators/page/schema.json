{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "page",
  "title": "Create an Analog page",
  "type": "object",
  "properties": {
    "pathname": {
      "type": "string",
      "description": "Name of the page to be generated. Ex. `products/[products]`, `products`, `blog/post`, `index`. __Note: special names like product.[productId] or (blog) should be surrounded by single quotes__",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-priority": "important",
      "x-prompt": "What name would you like to use for the page? Ex. `products/[products]`, `products`, `blog/post`, `index`"
    },
    "project": {
      "type": "string",
      "description": "The project where the page will be located.",
      "x-prompt": "Which project will this page be located in?",
      "$default": {
        "$source": "projectName"
      },
      "x-dropdown": "projects"
    },
    "metadata": {
      "description": "If metadata is needed for the route.",
      "type": "boolean",
      "default": false
    },
    "title": {
      "type": "string",
      "description": "The page title.",
      "default": "Page Title"
    },
    "redirectPage": {
      "description": "Whether the generated page is a redirect. _Note: Redirect path needs to be provided in the next prompt._",
      "type": "boolean",
      "default": false,
      "x-priority": "important",
      "x-prompt": "Is this a redirect page?"
    },
    "redirectPath": {
      "description": "The path to redirect to. _Note: This is only needed if the redirect page is set to true._",
      "type": "string",
      "default": "/",
      "x-priority": "important",
      "x-prompt": "What is the path to redirect to?"
    },
    "pathMatch": {
      "type": "string",
      "description": "The path to match for the redirect.",
      "default": "full",
      "enum": ["prefix", "full"],
      "x-prompt": "What is the path to match for the redirect?"
    }
  },
  "required": ["pathname", "project"]
}
