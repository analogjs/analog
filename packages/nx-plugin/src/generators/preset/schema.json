{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "Preset",
  "title": "",
  "type": "object",
  "properties": {
    "analogAppName": {
      "type": "string",
      "description": "",
      "x-prompt": "What name would you like to use for your Analog app?"
    },
    "addTailwind": {
      "type": "boolean",
      "description": "Adds tailwind if set to true.",
      "x-prompt": "Add TailwindCSS for styling?",
      "default": true
    },
    "addTRPC": {
      "type": "boolean",
      "description": "Adds tRPC if set to true.",
      "x-prompt": "Add tRPC for typesafe client/server interaction?",
      "default": false
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    }
  },
  "required": ["analogAppName"]
}
