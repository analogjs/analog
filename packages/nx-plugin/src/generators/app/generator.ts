import {
  formatFiles,
  getWorkspaceLayout,
  installPackagesTask,
  names,
  offsetFromRoot as determineOffsetFromRoot,
  stripIndents,
  Tree,
} from '@nx/devkit';
import { AnalogNxApplicationGeneratorOptions } from './schema';
import { major, coerce } from 'semver';
import { getInstalledPackageVersion } from '../../utils/version-utils';
import { addAnalogProjectConfig } from './lib/add-analog-project-config';
import { addAnalogDependencies } from './lib/add-analog-dependencies';
import { initializeAngularWorkspace } from './lib/initialize-analog-workspace';
import { addFiles } from './lib/add-files';
import { addTailwindConfig } from './lib/add-tailwind-config';
import { addTrpc } from './lib/add-trpc';
import { addHomePage } from './lib/add-home-page';
import {
  belowMinimumSupportedNxVersion,
  belowMinimumSupportedNxtRPCVersion,
} from './versions/minimum-supported-versions';

export interface NormalizedOptions
  extends AnalogNxApplicationGeneratorOptions,
    ReturnType<typeof names> {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
  offsetFromRoot: string;
  appsDir: string;
  nxPackageNamespace: string;
}

function normalizeOptions(
  tree: Tree,
  options: AnalogNxApplicationGeneratorOptions,
  nxVersion: string
): NormalizedOptions {
  const isNx = tree.exists('/nx.json');
  const appsDir = isNx ? getWorkspaceLayout(tree).appsDir : 'projects';
  const allNames = names(options.analogAppName);
  const projectDirectory = allNames.fileName;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];
  const offsetFromRoot = determineOffsetFromRoot(projectRoot);
  const nxPackageNamespace = major(nxVersion) >= 16 ? '@nx' : '@nrwl';
  const addTailwind = options.addTailwind ?? true;
  const addTRPC = options.addTRPC ?? false;

  return {
    ...options,
    ...allNames,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
    offsetFromRoot,
    appsDir,
    nxPackageNamespace,
    addTailwind,
    addTRPC,
  };
}

export async function appGenerator(
  tree: Tree,
  options: AnalogNxApplicationGeneratorOptions
) {
  const nxVersion = getInstalledPackageVersion(tree, 'nx');

  if (!nxVersion) {
    throw new Error(stripIndents`Nx must be installed to execute this plugin`);
  }

  if (belowMinimumSupportedNxVersion(nxVersion)) {
    throw new Error(
      stripIndents`Nx v15.2.0 or newer is required to install Analog`
    );
  }

  if (belowMinimumSupportedNxtRPCVersion(nxVersion) && options.addTRPC) {
    console.warn(
      'Nx v16.1.0 or newer is required to use tRPC with Analog. Skipping installation.'
    );
    options.addTRPC = false;
  }

  const normalizedOptions = normalizeOptions(tree, options, nxVersion);
  const angularVersion = await initializeAngularWorkspace(
    tree,
    nxVersion,
    normalizedOptions
  );
  const majorAngularVersion = major(coerce(angularVersion));

  await addAnalogDependencies(tree, nxVersion, angularVersion);

  const {
    projectRoot,
    projectName,
    parsedTags,
    name,
    appsDir,
    nxPackageNamespace,
  } = normalizedOptions;
  addAnalogProjectConfig(
    tree,
    projectRoot,
    projectName,
    parsedTags,
    name,
    appsDir,
    nxPackageNamespace
  );

  addFiles(tree, normalizedOptions, majorAngularVersion);

  if (normalizedOptions.addTailwind) {
    await addTailwindConfig(tree, projectName);
  }

  if (normalizedOptions.addTRPC) {
    await addTrpc(tree, projectRoot, nxVersion, normalizedOptions);
  }

  addHomePage(tree, normalizedOptions);

  // await addEslint(tree, majorNxVersion, normalizedOptions);

  if (!normalizedOptions.skipFormat) {
    await formatFiles(tree);
  }

  return () => {
    installPackagesTask(tree);
  };
}

export default appGenerator;
