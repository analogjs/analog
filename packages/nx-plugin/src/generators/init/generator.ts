import {
  formatFiles,
  generateFiles,
  getProjects,
  installPackagesTask,
  Tree,
} from '@nx/devkit';
import { join } from 'node:path';

import { getInstalledPackageVersion } from '../../utils/version-utils';
import { addAnalogDependencies } from './lib/add-analog-dependencies';
import { updateTestTsConfig } from './lib/update-test-tsconfig';
import { updateTestTarget } from './lib/update-test-target';
import { SetupAnalogGeneratorSchema } from './schema';
import { updateBuildTarget } from './lib/update-build-target';
import { updateServeTarget } from './lib/update-serve-target';
import { updatePackageJson } from './lib/update-package-json';
import { updateIndex } from './lib/update-index-html';
import { updateMain } from './lib/update-main';
import { updateAppTsConfig } from './lib/update-app-tsconfig';

function addFiles(tree: Tree, options: SetupAnalogGeneratorSchema) {
  const isNx = tree.read('/nx.json');
  const projects = getProjects(tree);
  const projectConfig = projects.get(options.project);

  const templateOptions = {
    ...options,
    offsetFromRoot: isNx ? '../../' : './',
    projectRoot: projectConfig.root,
    template: '',
  };

  generateFiles(
    tree,
    join(__dirname, 'files'),
    projectConfig.root || '.',
    templateOptions
  );

  if (options.vitest) {
    generateFiles(
      tree,
      join(__dirname, 'test-files'),
      projectConfig.root || '.',
      templateOptions
    );
  }
}

export async function setupAnalogGenerator(
  tree: Tree,
  options: SetupAnalogGeneratorSchema
) {
  const angularVersion = getInstalledPackageVersion(tree, '@angular/core');

  addAnalogDependencies(tree, angularVersion);
  updateBuildTarget(tree, options);
  updateServeTarget(tree, options);

  if (options.vitest) {
    updateTestTarget(tree, options);
    updateTestTsConfig(tree, options);
  }

  updateAppTsConfig(tree, options);
  updatePackageJson(tree, options);
  updateIndex(tree, options);
  updateMain(tree, options);

  addFiles(tree, options);

  await formatFiles(tree);

  return () => {
    installPackagesTask(tree);
  };
}

export default setupAnalogGenerator;
