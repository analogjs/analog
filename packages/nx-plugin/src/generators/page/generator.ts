import {
  convertNxGenerator,
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  stripIndents,
  Tree,
} from '@nx/devkit';
import { join } from 'node:path';
import { AnalogPageGeneratorSchema, NormalizedSchema } from './schema';

function normalizeOptions(
  tree: Tree,
  options: AnalogPageGeneratorSchema
): NormalizedSchema {
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${options.project}`;
  return {
    ...options,
    projectRoot,
  };
}

function generateFileName(input: string) {
  const pattern = /^[a-zA-Z0-9]+\.\[[a-zA-Z0-9-]+\]$/;
  if (pattern.test(input)) {
    return input.replace(/\[[a-zA-Z0-9-]+\]/, (match) => {
      const wordId = match.slice(1, -1);
      const camelCaseWordId = wordId.replace(/-([a-zA-Z0-9])/g, (_, letter) =>
        letter.toUpperCase()
      );
      return `[${camelCaseWordId}]`;
    });
  } else {
    return input;
  }
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const splitName = options.pathname.split('/');
  const routeName = splitName[splitName.length - 1];
  const fileName = generateFileName(routeName);
  const templateOptions = {
    ...options,
    ...names(routeName),
    name: names(routeName).fileName,
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    template: '',
    fileName,
  };

  const pageFolders = options.pathname.split('/').slice(0, -1);
  const pageDir = join(options.projectRoot, 'src/app/pages', ...pageFolders);

  generateFiles(tree, join(__dirname, 'files'), pageDir, templateOptions);
}

export async function analogPageGenerator(
  tree: Tree,
  options: AnalogPageGeneratorSchema
) {
  const normalizedOptions = normalizeOptions(tree, options);
  if (options.redirectPage && !options.redirectPath) {
    throw new Error(
      stripIndents`A redirectPath is required when redirectPage is true.`
    );
  }
  addFiles(tree, normalizedOptions);

  await formatFiles(tree);
}

export const analogPageGeneratorSchematic =
  convertNxGenerator(analogPageGenerator);

export default analogPageGenerator;
