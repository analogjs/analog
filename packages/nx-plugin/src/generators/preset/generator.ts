import { Tree } from '@nx/devkit';
import { PresetGeneratorSchema } from './schema';

export default async function (tree: Tree, options: PresetGeneratorSchema) {
  return await import('../app/generator').then(({ appGenerator }) =>
    appGenerator(tree, options)
  );
}
