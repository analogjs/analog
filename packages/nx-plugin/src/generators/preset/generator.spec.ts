import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import {
  readProjectConfiguration,
  addDependenciesToPackageJson,
} from '@nx/devkit';

import generator from './generator';
import { AnalogNxApplicationGeneratorOptions } from '../app/schema';

describe('preset generator', () => {
  const setup = async (
    options: AnalogNxApplicationGeneratorOptions,
    nxVersion = '17.0.0'
  ) => {
    const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    addDependenciesToPackageJson(tree, { nx: nxVersion }, {});
    await generator(tree, options);
    const config = readProjectConfiguration(tree, options.analogAppName);
    return {
      tree,
      config,
    };
  };

  it('should run successfully with latest', async () => {
    const { config } = await setup({ analogAppName: 'analog' });
    expect(config).toBeDefined();
  });

  it('should run successfully with Nx 16.x', async () => {
    const { config } = await setup({ analogAppName: 'analog' }, '16.10.0');
    expect(config).toBeDefined();
  });
});
