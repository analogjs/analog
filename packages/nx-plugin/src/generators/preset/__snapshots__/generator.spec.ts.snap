// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preset generator > should match index.html 1`] = `
"<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>my-app</title>
    <base href="/" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link rel="icon" type="image/x-icon" href="/favicon.ico" />
  </head>
  <body>
    <app-root></app-root>
    <script type="module" src="/my-app/src/main.ts"></script>
  </body>
</html>
"
`;

exports[`preset generator > should match project.json 1`] = `
"{
  "name": "my-app",
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "projectType": "application",
  "prefix": "app",
  "sourceRoot": "my-app/src",
  "tags": ["tag1", "tag2", "type:app"],
  "targets": {
    "build": {
      "executor": "@analogjs/platform:vite",
      "options": {
        "configFile": "my-app/vite.config.ts",
        "main": "my-app/src/main.ts",
        "outputPath": "dist/my-app/client",
        "tsConfig": "my-app/tsconfig.app.json"
      },
      "defaultConfiguration": "production",
      "configurations": {
        "development": {
          "mode": "development"
        },
        "production": {
          "sourcemap": false,
          "mode": "production"
        }
      },
      "outputs": ["{workspaceRoot}/dist/apps/my-app"]
    },
    "serve": {
      "executor": "@analogjs/platform:vite-dev-server",
      "defaultConfiguration": "development",
      "options": {
        "buildTarget": "my-app:build",
        "port": 4200
      },
      "configurations": {
        "development": {
          "buildTarget": "my-app:build:development",
          "hmr": true
        },
        "production": {
          "buildTarget": "my-app:build:production"
        }
      }
    },
    "extract-i18n": {
      "executor": "@angular-devkit/build-angular:extract-i18n",
      "options": {
        "buildTarget": "my-app:build"
      }
    },
    "test": {
      "executor": "@analogjs/vitest-angular:test",
      "options": {
        "config": "vite.config.ts"
      }
    },
    "serve-static": {
      "continuous": true,
      "executor": "@nx/web:file-server",
      "options": {
        "buildTarget": "my-app:build",
        "port": 4200,
        "staticFilePath": "dist/my-app/browser",
        "spa": true
      }
    }
  }
}
"
`;

exports[`preset generator > should match src/test-setup.ts 1`] = `
"import '@analogjs/vitest-angular/setup-zone';

import {
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting,
} from '@angular/platform-browser-dynamic/testing';
import { getTestBed } from '@angular/core/testing';

getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting(),
);
"
`;

exports[`preset generator > should match vite.config.ts 1`] = `
"/// <reference types="vitest" />

import analog from '@analogjs/platform';
import { defineConfig } from 'vite';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

// @ts-expect-error @tailwindcss/vite currently uses mts. TypeScript is complaining this, but it works as expected.
import tailwindcss from '@tailwindcss/vite';

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => {
  return {
    root: __dirname,
    cacheDir: \`../node_modules/.vite\`,
    build: {
      outDir: '../dist/./my-app/client',
      reportCompressedSize: true,
      target: ['es2020'],
    },
    server: {
      fs: {
        allow: ['.'],
      },
    },
    plugins: [tailwindcss(), analog(), nxViteTsPaths()],
    test: {
      globals: true,
      environment: 'jsdom',
      setupFiles: ['src/test-setup.ts'],
      include: ['**/*.spec.ts'],
      reporters: ['default'],
    },
    define: {
      'import.meta.vitest': mode !== 'production',
    },
  };
});
"
`;

exports[`preset generator > should match vite.config.ts 2`] = `
"/// <reference types="vitest" />

import analog from '@analogjs/platform';
import { defineConfig } from 'vite';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

// @ts-expect-error @tailwindcss/vite currently uses mts. TypeScript is complaining this, but it works as expected.
import tailwindcss from '@tailwindcss/vite';

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => {
  return {
    root: __dirname,
    cacheDir: \`../node_modules/.vite\`,
    build: {
      outDir: '../dist/./my-app/client',
      reportCompressedSize: true,
      target: ['es2020'],
    },
    server: {
      fs: {
        allow: ['.'],
      },
    },
    plugins: [tailwindcss(), analog(), nxViteTsPaths()],
    test: {
      globals: true,
      environment: 'jsdom',
      setupFiles: ['src/test-setup.ts'],
      include: ['**/*.spec.ts'],
      reporters: ['default'],
    },
    define: {
      'import.meta.vitest': mode !== 'production',
    },
  };
});
"
`;
